include/swift/AST/TypeCheckRequests.h

```diff
@@ -78,13 +79,63 @@ void simple_display(
 
 void simple_display(llvm::raw_ostream &out, ASTContext *ctx);
 
+/// Emulates the following enum with associated values:
+/// enum InheritedTypeResult {
+///     case inherited(Type)
+///     case SuppressedInference(KnownProtocolKind)
+///     case `default`
+/// }
+class InheritedTypeResult {
```

Comment:
This will introduce a second representation for the `~Protocol` syntax in the AST: For the "invertible" protocols like `Copyable`, we represent them as a `ProtocolCompositionType`, which has its own list of inverses. I think that representation doesn't make sense for this use case, so it's okay to have two.